cpp/src_prims/sparse/csr.h

```diff
@@ -47,170 +47,204 @@ static const float MIN_FLOAT = std::numeric_limits<float>::min();
  * while the original caller still maintains ownership of the underlying memory.
  *
  * @tparam T: the type of the value array.
+ * @tparam Index_Type: The type of the index arrays
  *
  */
-template <typename T>
+
+template <typename T, typename Index_Type = int>
 class CSR {
+ protected:
+  device_buffer<Index_Type> row_ind;
+  device_buffer<Index_Type> row_ind_ptr;
+  device_buffer<T> vals;
+
  public:
-  int *row_ind;
-  int *row_ind_ptr;
-  T *vals;
-  int nnz;
-  int n_rows;
-  int n_cols;
+  Index_Type nnz;
+  Index_Type n_rows;
+  Index_Type n_cols;
 
   /**
-     * @brief default constructor
-     */
-  CSR()
-    : row_ind(nullptr),
-      row_ind_ptr(nullptr),
-      vals(nullptr),
-      nnz(-1),
-      n_rows(-1),
-      n_cols(-1) {}
-
-  /*
-     * @brief construct a CSR object with arrays
-     *
-     * @param row_ind: the array of row_indices
-     * @param row_ind_ptr: array of row_index_ptr
-     * @param vals: array of data
-     * @param nnz: size of data and row_ind_ptr arrays
-     * @param n_rows: number of rows in the dense matrix
-     * @param n_cols: number of cols in the dense matrix
-     */
-  CSR(int *row_ind, int *row_ind_ptr, T *vals, int nnz, int n_rows = -1,
-      int n_cols = -1) {
-    this->row_ind = row_ind;
-    this->row_ind_ptr = row_ind_ptr;
-    this->vals = vals;
-    this->nnz = nnz;
-    this->n_rows = n_rows;
-    this->n_cols = n_cols;
-  }
-  /*
-     * @brief construct an empty allocated CSR given its size
-     *
-     * @param nnz: size of data and row_ind_ptr arrays
-     * @param n_rows: number of rows in the dense matrix
-     * @param n_cols: number of cols in the dense matrix
-     * @param init: initialize arrays to zeros?
-     */
-
-  CSR(int nnz, int n_rows = -1, int n_cols = -1, bool init = true)
-    : row_ind(nullptr),
-      row_ind_ptr(nullptr),
-      vals(nullptr),
+    * @param device: are the underlying arrays going to be on device?
+    */
+  CSR(std::shared_ptr<deviceAllocator> alloc, cudaStream_t stream)
+    : row_ind(alloc, stream, 0),
+      row_ind_ptr(alloc, stream, 0),
+      vals(alloc, stream, 0),
+      nnz(0),
+      n_rows(0),
+      n_cols(0) {}
+
+  /**
+    * @param row_ind_: csr row index array
+    * @param row_ind_ptr_: csr row index pointer array
+    * @param vals: csr vals array
+    * @param nnz: size of the rows/cols/vals arrays
+    * @param n_rows: number of rows in the dense matrix
+    * @param n_cols: number of cols in the dense matrix
+    */
+  CSR(device_buffer<Index_Type> &row_ind_,
+      device_buffer<Index_Type> &row_ind_ptr_, device_buffer<T> &vals,
+      Index_Type nnz, Index_Type n_rows = -1, Index_Type n_cols = -1)
```

Comment:
default `n_rows` and `n_cols` here is -1, whereas in the default ctor, they have been updated to 0. Is this deliberate?